### PR TITLE
Add code to generate UUID for one of the HMRC headers #41

### PR DESCRIPTION
**Pull request by salilbc**
# Title
Add code to generate UUID for one of the HMRC headers #41

## Description of what's changing
Added a new Header Gov-Client-Device-ID
...

## What else might be impacted?

...

## Which issue does this PR relate to?
#41 
...

## Checklist

[yes] Tests are written and maintain or improve code coverage
[yes] I've tested this in my application using `yarn link` (if applicable)
[yes ] You consent to and are confident this change can be released with no regression

### CODE REVIEW

File: src/js/common/browserInfoHelper.js [hunk 1]
```diff
@@ -106,3 +107,7 @@ export const getBrowserDoNotTrackStatus = () => {
       windowVar.external.msTrackingProtectionEnabled());
   return isBrowserDoNotTrack ? "true" : "false";
 };
+
+export const getDeviceId = () => {
+  return uuid();
```
Comment (reubenae) [inline src/js/common/browserInfoHelper.js:112]: I don't think this is the right location. UUID generation is nothing to do with getting data out of the browser API's. Suggest its own helper file or no util file at all.

File: tests/unit/common/browserInfoHelper.test.js [hunk 2]
```diff
@@ -259,4 +260,10 @@ describe("BrowserInfoHelper", () => {
     expect(getScreenHeight()).toEqual(1021);
     expect(getScreenColourDepth()).toEqual(17);
   });
+
+  it("getDeviceId", () => {
+    const getUUID = getDeviceId();
+    expect(getUUID.split("-").length).toEqual(5);
```
Comment (reubenae) [inline tests/unit/common/browserInfoHelper.test.js:266]: Please see https://github.com/intuit/user-data-for-fraud-prevention/pull/52#discussion_r498529292 for my thoughts on this testing strategy.

**Review by reubenae — CHANGES_REQUESTED** [review]

File: tests/unit/hmrc/mtdFraudPrevention.test.js [hunk 3]
```diff
@@ -93,6 +102,9 @@ describe("FraudPreventionHeaders", () => {
     );
     expect(headers.get("Gov-Client-Browser-Do-Not-Track")).toBe("true");
     expect(headers.get("Gov-Client-Local-IPs")).toBe(undefined);
+    expect(headers.get("Gov-Client-Device-ID")).toBe(
+        "uuid=424f48aa-b723-4f97-8a30-d214b43bf372"
```
Comment (reubenae) [inline tests/unit/hmrc/mtdFraudPrevention.test.js:106]: As per the issue, the expected format of these headers is `Gov-Client-Device-ID: beec798b-b366-47fa-b1f8-92cede14a1ce`

Please refer to the documentation on how to run your tests locally

**Review by reubenae — CHANGES_REQUESTED** [review]

File: src/js/common/browserInfoHelper.js (hunk 1 above)
Comment (salilbc) [reply src/js/common/browserInfoHelper.js]: resolved

File: tests/unit/common/browserInfoHelper.test.js (hunk 2 above)
Comment (salilbc) [reply tests/unit/common/browserInfoHelper.test.js]: sure

File: tests/unit/hmrc/mtdFraudPrevention.test.js (hunk 3 above)
Comment (salilbc) [reply tests/unit/hmrc/mtdFraudPrevention.test.js]: updated the code to reflect the changes

File: tests/unit/common/uuidGeneratorHelper.test.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { getDeviceId } from "../../../src/js/common/uuidGeneratorHelper";
+import * as uuidGeneratorHelper from "../../../src/js/common/uuidGeneratorHelper";
+
+describe("uuidGeneratorHelper", () => {
+    afterEach(() => {
+        jest.resetAllMocks();
+    });
+
+    it("getDeviceId", () => {
+        jest.spyOn(uuidGeneratorHelper, "getDeviceId").mockReturnValue("424f48aa-b723-4f97-8a30-d214b43bf372");
```
Comment (reubenae) [inline tests/unit/common/uuidGeneratorHelper.test.js:10]: This test doesn't really do anything. You are asserting that your spy returns the value you tell it to return. You should stub/mock/spy `uuid` package itself

File: tests/unit/hmrc/mtdFraudPrevention.test.js [hunk 5]
```diff
@@ -35,6 +38,7 @@ describe("FraudPreventionHeaders", () => {
       height: 1021,
       colorDepth: 17,
     });
+    jest.spyOn(uuidGeneratorHelper, "getDeviceId").mockReturnValue("af33f46c-085d-4050-be2a-98b72820c572");
```
Comment (reubenae) [inline tests/unit/hmrc/mtdFraudPrevention.test.js:41]: Same again, this test class is an integration test, i.e. it should not matter if we modify helper files, restructure our project. Mock external dependencies (`uuid` in this case) and test the end to end solution. A good thought experiment here to know if you are doing the right test: can you restructure your internal implementation without your test breaking? For example, if you were to just rename `getDeviceId` to be something else, your test will break... we don't want that! That is way too brittle.

File: src/js/common/uuidGeneratorHelper.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import uuid from "uuid";
+
+export const getDeviceId = () => {
```
Comment (reubenae) [inline src/js/common/uuidGeneratorHelper.js:3]: Please change this to be like this:

`export const getDeviceId = () => uuid();`

File: src/js/common/uuidGeneratorHelper.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import uuid from "uuid";
```
Comment (reubenae) [inline src/js/common/uuidGeneratorHelper.js:1]: I'm not sure I like this file name. What else will ever fit into "uuidGeneratorHelper"? I suggest you either make this name much more generic, or, my preference, just use uuid where you actually need it (in mtdFraudPrevention.js)

Abstracting this function out into another file is just going to make reading through the code more difficult as you have to jump around IMO. However if you can come up with a good name for this utility file such that future engineers will probably add to it, I'm okay with that approach too.

**Review by reubenae — CHANGES_REQUESTED** [review]

File: src/js/common/uuidGeneratorHelper.js (hunk 7 above)
Comment (salilbc) [reply src/js/common/uuidGeneratorHelper.js]: moved it to mtdFraudPrevention

File: src/js/common/uuidGeneratorHelper.js (hunk 6 above)
Comment (salilbc) [reply src/js/common/uuidGeneratorHelper.js]: done

File: src/js/hmrc/mtdFraudPrevention.js [hunk 8]
```diff
@@ -59,6 +62,10 @@ export const getFraudPreventionHeaders = async () => {
       header: fraudPreventionHeadersEnum.DEVICE_LOCAL_IPS,
       callback: async () => encodeURI(await getDeviceLocalIPAsString()),
     },
+    {
+      header: fraudPreventionHeadersEnum.DEVICE_ID,
+      callback: getDeviceId,
```
Comment (reubenae) [inline src/js/hmrc/mtdFraudPrevention.js:67]: I guess this can just be `callback: uuid,`

Comment (salilbc) [reply src/js/hmrc/mtdFraudPrevention.js]: done

File: tests/unit/hmrc/mtdFraudPrevention.test.js (hunk 5 above)
Comment (salilbc) [reply tests/unit/hmrc/mtdFraudPrevention.test.js]: done

File: tests/unit/common/uuidGeneratorHelper.test.js (hunk 4 above)
Comment (salilbc) [reply tests/unit/common/uuidGeneratorHelper.test.js]: done

File: tests/unit/hmrc/mtdFraudPrevention.test.js (hunk 5 above)
Comment (salilbc) [reply tests/unit/hmrc/mtdFraudPrevention.test.js]: removed dependency on getDeviceId

**Comment by skodamarthi** [comment]
hi @salilbc! We really appreciate your contribution! The code changes look good. However since the file structure on https://github.com/intuit/user-data-for-fraud-prevention/pull/52 was what we were expecting (to have common code in the common folder), we went ahead with merging that PR. Thanks a lot for your contribution!

**Pull request closed without merging**